src/System.Management.Automation/engine/runtime/Operations/ClassOps.cs

```diff
@@ -43,9 +43,26 @@ internal SessionStateKeeper()
 
         internal void RegisterRunspace()
         {
-            // it's not get, but really 'Add' value.
-            // ConditionalWeakTable.Add throw exception, when you are trying to add a value with the same key.
-            _stateMap.GetValue(Runspace.DefaultRunspace, runspace => runspace.ExecutionContext.EngineSessionState);
+            SessionStateInternal ssInMap = null;
+            Runspace rsToUse = Runspace.DefaultRunspace;
+            SessionStateInternal ssToUse = rsToUse.ExecutionContext.EngineSessionState;
+            
+            // Different threads will operate on different key/value pairs (default-runspace/session-state pairs),
+            // and a ConditionalWeakTable itself is thread safe, so there won't be race condition here.
+            if (!_stateMap.TryGetValue(rsToUse, out ssInMap))
+            {
+                // If the key doesn't exist yet, add it
+                _stateMap.Add(rsToUse, ssToUse);
+            }
+            else if (!ssInMap.Equals(ssToUse))
+            {
+                // If the key exists but the corresponding value is not what we should use, then remove the key/value pair and add the new pair.
+                // This could happen when a powershell class is defined in a module and the module gets reloaded. In such case, the same TypeDefinitionAst
+                // instance will get reused, but should be associated with the SessionState from the new module, instead of the one from the old module.
+                _stateMap.Remove(rsToUse);
```

Comment:
For existing class instances, I think they will continue to run against the original session state because the session state to be used is stored in `_sessionStateField` when an instance is constructed, see the code [here](https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/parser/PSType.cs#L801).

Here is what I got in such scenario:
**test.psm1**
```
$passedArgs = $args
function Get-PassedValue { $passedArgs }
class Root
{
    $property = $passedArgs
}
function Get-Root { [Root]::new() }
```

Run existing class instance after updating the `SessionStateKeepper` map
```
PS F:\tmp> Import-Module .\test.psm1 -ArgumentList 'value1'
PS F:\tmp> $root = Get-Root
PS F:\tmp>
PS F:\tmp> $root.property
value1
PS F:\tmp>
PS F:\tmp> Import-Module .\test.psm1 -ArgumentList 'value2' -Force
PS F:\tmp>
PS F:\tmp> $soot = Get-Root
PS F:\tmp>
PS F:\tmp> $soot.property
value2
PS F:\tmp> $root.property
value1
```